anxcloud/resource_virtual_server.go

```diff
@@ -290,33 +289,17 @@ func resourceVirtualServerRead(ctx context.Context, d *schema.ResourceData, m in
 		return nil
 	}
 
-	err = resource.RetryContext(ctx, d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
-		info, infoErr := vsphereAPI.Info().Get(ctx, d.Id())
-		if infoErr != nil {
-			if err := handleNotFoundError(infoErr); err != nil {
-				return resource.NonRetryableError(fmt.Errorf("unable to get vm with id '%s': %w", d.Id(), err))
-			}
-			d.SetId("")
-			return nil
-		}
-		var nicErr error
-		for _, nic := range info.Network {
-			if len(nic.IPv4) == 0 && len(nic.IPv6) == 0 {
-				nicErr = multierror.Append(nicErr, fmt.Errorf("missing IPs for NIC"))
-			}
-		}
-		if nicErr != nil {
-			return resource.RetryableError(nicErr)
-		}
-		return nil
-	})
+	nicTypes, err := nicAPI.List(ctx)
 	if err != nil {
 		return diag.FromErr(err)
 	}
 
-	nicTypes, err := nicAPI.List(ctx)
-	if err != nil {
-		return diag.FromErr(err)
+	// from_scratch template_id's aren't set at the info endpoint
+	// that's why we keep the configured template_id if not returned by API
+	if info.TemplateID != "" {
+		if err = d.Set("template_id", info.TemplateID); err != nil {
+			diags = append(diags, diag.FromErr(err)...)
+		}
```

Comment:
should we limit this code to VMs that are `from_scratch`?

Reply:
I guess you mean, that the `template_id` should be set, only if the `template_type` is not `from_scratch` like the following?
```suggestion
	// from_scratch template_id's aren't set at the info endpoint
	// that's why we keep the configured template_id if not returned by API
	if templateType, ok := d.Get("template_type").(string); !ok || templateType != "from_scratch" {
		if err = d.Set("template_id", info.TemplateID); err != nil {
			diags = append(diags, diag.FromErr(err)...)
		}
```

Reply:
more like `if from_scratch { allow info.TemplateID not being set } else { error on info.TemplateID not set }`

but that's what this code is already doing, right? Except not checking if TemplateID is empty, but at least always storing the TemplateID given by the vminfo API in our terraform state

so yeah, think this works :)

Reply:
Correct, the current implementation skips updating the `template_id` on read when `template_type` is "from_scratch" to preserve the configured id. I think an additional check if the `template_id` in the response is unnecessary, as it is never set for this template type.